File: apps/web/src/lib/social/fundraiser_progress.ts
/**
 * Fundraiser thermometer math (fundraising.md).
 *
 * Pure, deterministic, no I/O — computes the goal-thermometer state from
 * `(raisedCents, goalCents)`: the bar fill percentage (clamped to 0–100 for
 * the bar geometry), the true uncapped percentage (so the UI can show
 * "118% — over goal!"), the cents still needed, and a coarse `ThermometerState`
 * that drives the bar styling + label.
 *
 * TS↔Dart parity pair: keep in lockstep with
 * `apps/mobile_android/lib/fundraiser_progress.dart` (+ iOS twin) — same
 * algorithm, same edge cases, same test count.
 */

export type ThermometerState = 'starting' | 'progressing' | 'met' | 'exceeded';

export interface FundraiserProgress {
	/** Bar fill 0–100 (clamped), for the thermometer geometry. */
	fillPct: number;
	/** True percentage of goal raised, uncapped (can exceed 100). */
	rawPct: number;
	/** Cents still needed to reach the goal; 0 once met or exceeded. */
	remainingCents: number;
	state: ThermometerState;
}

/** Below this fill the bar reads as "just getting started". */
export const STARTING_THRESHOLD_PCT = 10;

/**
 * Resolve the thermometer state from raised vs goal.
 *
 * Defensive: a non-finite or non-positive goal yields a zeroed,
 * `'starting'` result (the bar renders empty rather than dividing by zero);
 * negative raised is floored to 0.
 */
export function fundraiserProgress(
	raisedCents: number,
	goalCents: number,
): FundraiserProgress {
	const raised = Number.isFinite(raisedCents) && raisedCents > 0 ? raisedCents : 0;
	if (!Number.isFinite(goalCents) || goalCents <= 0) {
		return { fillPct: 0, rawPct: 0, remainingCents: 0, state: 'starting' };
	}

	const rawPct = (raised / goalCents) * 100;
	const fillPct = Math.max(0, Math.min(100, rawPct));
	const remainingCents = Math.max(0, goalCents - raised);

	let state: ThermometerState;
	if (raised > goalCents) {
		state = 'exceeded';
	} else if (raised >= goalCents) {
		state = 'met';
	} else if (rawPct < STARTING_THRESHOLD_PCT) {
		state = 'starting';
	} else {
		state = 'progressing';
	}

	return { fillPct, rawPct, remainingCents, state };
}

File: apps/mobile_android/lib/fundraiser_progress.dart
/// Fundraiser thermometer math (fundraising.md).
///
/// Pure, deterministic, no I/O — computes the goal-thermometer state from
/// `(raisedCents, goalCents)`: the bar fill percentage (clamped to 0–100 for
/// the bar geometry), the true uncapped percentage (so the UI can show
/// "118% — over goal!"), the cents still needed, and a coarse [ThermometerState]
/// that drives the bar styling + label.
///
/// TS↔Dart parity pair: keep in lockstep with
/// `apps/web/src/lib/social/fundraiser_progress.ts` — same algorithm, same
/// edge cases, same test count.
library;

enum ThermometerState { starting, progressing, met, exceeded }

class FundraiserProgress {
  /// Bar fill 0–100 (clamped), for the thermometer geometry.
  final double fillPct;

  /// True percentage of goal raised, uncapped (can exceed 100).
  final double rawPct;

  /// Cents still needed to reach the goal; 0 once met or exceeded.
  final int remainingCents;

  final ThermometerState state;

  const FundraiserProgress({
    required this.fillPct,
    required this.rawPct,
    required this.remainingCents,
    required this.state,
  });
}

/// Below this fill the bar reads as "just getting started".
const double startingThresholdPct = 10;

/// Resolve the thermometer state from raised vs goal.
///
/// Defensive: a non-finite or non-positive goal yields a zeroed,
/// [ThermometerState.starting] result (the bar renders empty rather than
/// dividing by zero); negative raised is floored to 0.
FundraiserProgress fundraiserProgress(num raisedCents, num goalCents) {
  final raised = raisedCents.isFinite && raisedCents > 0 ? raisedCents : 0;
  if (!goalCents.isFinite || goalCents <= 0) {
    return const FundraiserProgress(
      fillPct: 0,
      rawPct: 0,
      remainingCents: 0,
      state: ThermometerState.starting,
    );
  }

  final rawPct = (raised / goalCents) * 100;
  final fillPct = rawPct.clamp(0, 100).toDouble();
  final remainingCents = (goalCents - raised) > 0 ? (goalCents - raised).round() : 0;

  final ThermometerState state;
  if (raised > goalCents) {
    state = ThermometerState.exceeded;
  } else if (raised >= goalCents) {
    state = ThermometerState.met;
  } else if (rawPct < startingThresholdPct) {
    state = ThermometerState.starting;
  } else {
    state = ThermometerState.progressing;
  }

  return FundraiserProgress(
    fillPct: fillPct,
    rawPct: rawPct.toDouble(),
    remainingCents: remainingCents,
    state: state,
  );
}
